src/datatrove/io.py

```diff
@@ -302,4 +303,21 @@ def download_file(remote_path: str, local_path: str, progress: bool = True):
     )
 
 
+def download_file_safely(remote_path: str, local_path: str, progress: bool = True):
+    """
+    Download a file from a remote path to a local path.
+    This function is process-safe and will only download the file if it hasn't been downloaded already.
+    Args:
+        remote_path: str: The remote path to the file to download
+        local_path: str: The local path to save the file to
+        progress: bool: Whether to show a progress bar (Default value = True)
+    """
+    with InterProcessLock(f"{local_path}.lock"):
+        # Make sure to do exists check process-locked as otherwise race condition can happen, and processes,
+        # which don't perform the download, could try to open the file before it's fully downloaded
+        if os.path.exists(local_path):
```

Comment:
How much does this slow down jobs when the file has already been downloaded previously and we have many tasks? Is there maybe some other smart combination of checks we could do (like check if the file to download exists AND the lock file does not exist) before creating a lock?

Reply:
Tested the following pipeline on 1k tasks:
```
        [
                lambda x,y,z: [Document("","", metadata={"url": "a.cz"})],
                URLFilter(),
        ],
```

No interprocess-guard:
```
real    1m35.601s
user    0m11.359s
sys     0m5.849s
```

With interprocess-guard:
```
real    2m29.377s
user    0m11.822s
sys     0m3.853s
```

So I implemented your idea. Bear in mind that it's not 100% bulettproof as I check for existence of lock using simple fs check instead of syscall (cause win don't support it and fnctl support in python is scuffed)

Reply:
thought of a potentially safer approach: instead of checking if `file` exists, we can check if `file.processed` (or `file.downloaded`, something like this) exists, which is an empty file created only after the download is completed.

Reply:
The current approach will work, there will be no race condition/ deadlock whatsoever. The only problem is that if the lock is not deleted, all processes will have to sequentially in this section instead of skipping it completely and once they do and remove the lock from fs, the next run will be able to skip the section completely. So this bad situation happens only if we crash and only the first run after the crash.